### PR TITLE
add notification service

### PR DESCRIPTION
**Pull request by DarPers**

### CODE REVIEW

**Review by GodlyCat — APPROVED** [review]

File: NotificationService/NotificationService/Constants/SettingConstants.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿namespace NotificationService.Constants;
+public static class SettingConstants
```
Comment (nantonov) [inline NotificationService/NotificationService/Constants/SettingConstants.cs:2]: Move to appsettings?

**Review by nantonov — CHANGES_REQUESTED** [review]

File: NotificationService/NotificationService/EmailSender.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+public static class EmailSender
+{
+    public static async Task SendEmail(string email, string subject, string message)
+    {
+        var mail = SettingConstants.MainMail;
+
+        try
+        {
+            using (SmtpClient client = new SmtpClient())
```
Comment (nantonov) [inline NotificationService/NotificationService/EmailSender.cs:14]: Can we move it to DI?

File: NotificationService/NotificationService/EmailSender.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            {
+                client.Host = SettingConstants.SmtpHostName;
+                client.Credentials = new System.Net.NetworkCredential(mail, SettingConstants.PasswordForApp);
+                client.Port = 587;
```
Comment (nantonov) [inline NotificationService/NotificationService/EmailSender.cs:18]: Move it to appsettings?

File: NotificationService/NotificationService/Constants/SettingConstants.cs (hunk 1 above)
Comment (DarPers) [reply NotificationService/NotificationService/Constants/SettingConstants.cs:2]: This is part of console app, so there is no appsetting file

File: NotificationService/NotificationService/MessageHandler.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using NotificationService.Constants;
+using ServiceDesk.BLL.Interfaces;
+
+namespace NotificationService;
+public static class MessageHandler
+{
+    public static async Task HandleMessageAddedTicket(ITicketAdded message)
+    {
+        Console.WriteLine(MessageConstants.LogNewTicketMessage(message.CommandId));
```
Comment (nantonov) [inline NotificationService/NotificationService/MessageHandler.cs:9]: Logger?

File: NotificationService/NotificationService/MessageHandler.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        foreach (var receiver in message.ExecutorsEmails)
+        {
+            await EmailSender.SendEmail(receiver, MessageConstants.SubjectForNewTicket, message.Message);
```
Comment (nantonov) [inline NotificationService/NotificationService/MessageHandler.cs:12]: can we improve the performance here? 
await in the loop is not the proper way

File: NotificationService/NotificationService/MessageHandler.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+        }
+    }
+
+    public static Task HandleMessageUpdatedTicketStatus(ITicketStatusChanged message)
+    {
+        Console.WriteLine(MessageConstants.LogUpdatedTicketMessage(message.CommandId));
```
Comment (nantonov) [inline NotificationService/NotificationService/MessageHandler.cs:19]: Logger?

File: NotificationService/NotificationService/Program.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using NotificationService;
+using NotificationService.Constants;
+
+var bus = BusConfigureManager.SetUpBus();
+
+await bus.StartAsync();
+
+Console.WriteLine(MessageConstants.BusListeningMessage);
+
+Console.ReadKey();
```
Comment (nantonov) [inline NotificationService/NotificationService/Program.cs:10]: Can we add an exit possibility only if "q" was pressed?

File: ServiceDesk.API/ServiceDesk.BLL/Constants/MessageConstants.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    public static string CreateMessageAboutChangedStatus(Ticket ticket)
+    {
+        switch (ticket.Status)
+        {
+            case Status.InProgress:
+                return CreateTicketMessageWithName(ticket.Name) + "in progress!";
+            case Status.Ready:
+                return CreateTicketMessageWithName(ticket.Name) + "is ready!";
+            default:
+                return CreateTicketMessageWithName(ticket.Name) + $"has {ticket.Status} status now!";
```
Comment (nantonov) [inline ServiceDesk.API/ServiceDesk.BLL/Constants/MessageConstants.cs:29]: exception should be thrown?

File: ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs [hunk 9]
```diff
@@ -37,8 +39,13 @@ public async Task<TicketModel> SetTicketStatus(Guid id, Status newStatus, Cancel
 
         EntityIsNullException.ThrowIfNull(ticket);
 
-        ticket.Status = newStatus;
+        ticket!.Status = newStatus;
```
Comment (nantonov) [inline ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs:42]: It's better to check null value

File: ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs [hunk 10]
```diff
@@ -37,8 +39,13 @@ public async Task<TicketModel> SetTicketStatus(Guid id, Status newStatus, Cancel
...
         await _ticketRepository.UpdateEntityAsync(ticket, cancellationToken);
+
+        var user = await _userRepository.GetEntityByIdAsync(ticket.UserId, cancellationToken);
+
+        await _massageService.SendCommandChangedStatus(ticket, user!.Email);
```
Comment (nantonov) [inline ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs:47]: again, check null value instead

**Review by nantonov — CHANGES_REQUESTED** [review]

File: NotificationService/NotificationService/Constants/SettingConstants.cs (hunk 1 above)
Comment (nantonov) [reply NotificationService/NotificationService/Constants/SettingConstants.cs:2]: you can add

File: ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs (hunk 9 above)
Comment (DarPers) [reply ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs:42]: Checking null value is above

File: ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs (hunk 10 above)
Comment (DarPers) [reply ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs:47]: Checking null value is above

File: ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs (hunk 9 above)
Comment (nantonov) [reply ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs:42]: In this case we can delete '!'?

File: ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs (hunk 10 above)
Comment (nantonov) [reply ServiceDesk.API/ServiceDesk.BLL/Services/TicketService.cs:47]: In this case we can delete '!'?

**Review by nantonov — APPROVED** [review]

**Review by VadimZharikov — APPROVED** [review]

**Pull request merged**